## 🧬 Behavioral Analytics: The Modern Student Triad
### Exploring AI Dependency, Career Anxiety, and Student Burnout

### 🧠 Project Motivation
Most publicly available education datasets focus strictly on traditional performance metrics like grades, test scores, or attendance records. While useful, they completely miss the massive psychological shifts occurring in the generative AI era. 

This notebook explores the intersection of three highly correlated contemporary phenomena: **Generative AI Dependency**, **Acute Career Anxiety**, and **Chronic Student Burnout**. Rather than existing as isolated issues, these factors form a compounding psychological feedback loop that alters how students learn, cope, and view their professional futures.

### 🛠️ Dataset Design Philosophy
Unlike simplistic synthetic datasets with perfectly linear relationships, this dataset was intentionally engineered to include real-world dataset characteristics:
* **Overlapping Behaviors:** Students exhibiting high AI usage alongside high manual study efforts.
* **Contradictory Personas:** High-performing students tracking high burnout levels.
* **Controlled Missingness:** Simulating real-world survey drops or unrecorded tracking metrics.
* **Moderate Correlations:** Introducing realistic noise and variance to prevent trivial decision boundaries.

### Automated Data Ingestion
To ensure maximum reproducibility and maintain clean versioning, we pull the official dataset directly from Kaggle utilizing the `kagglehub` library. 

This automated process fetches the raw behavioral data tracking AI usage metrics, burnout indicators, and career anxiety scales directly into our environment.

* **Dataset Credit:** Sridip Basu via Kaggle (*AI Dependency, Career Anxiety, and Student Burnout Dataset*).

**Link:** https://www.kaggle.com/datasets/sridipbasu/ai-depndency-career-anxiety-and-student-burnout


### 📥 Loading the Dataset and Libraries

Before we start, we need to install the necessary Python libraries and **load the dataset**.


In [36]:
# Install the required libraries
%pip install pandas numpy matplotlib seaborn scikit-learn scikit-optimize kagglehub jupyterlab nbconvert imblearn xgboost sweetviz joblib -q

# Install and update the watermark package to display environment and library version information
%pip install -q -U watermark

Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.


In [37]:
# ==========================================
# 1. DATA MANIPULATION & VISUALIZATION
# ==========================================
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import joblib
import sweetviz as sv
import os
import kagglehub
import logging
import warnings
from sweetviz import FeatureConfig

warnings.filterwarnings('ignore', category=UserWarning)
logging.getLogger("matplotlib.font_manager").setLevel(logging.ERROR)

# ==========================================
# 2. PREPROCESSING & PIPELINES
# ==========================================
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder

# ==========================================
# 3. MACHINE LEARNING MODELS
# ==========================================
from sklearn.dummy import DummyClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from skopt import BayesSearchCV
from skopt.space import Real, Categorical

# ==========================================
# 4. MODEL SELECTION & HYPERTUNING
# ==========================================
from sklearn.model_selection import (
    train_test_split,
    StratifiedKFold,
    cross_val_score,
    cross_val_predict,
    GridSearchCV,
)

# ==========================================
# 5. METRICS & EVALUATION
# ==========================================
from sklearn.metrics import (
    classification_report,
    confusion_matrix,
    ConfusionMatrixDisplay,
)
from sklearn.metrics import roc_curve, auc
from sklearn.metrics import precision_recall_curve, average_precision_score

In [38]:
# Load the watermark extension to log the environment state
%reload_ext watermark

# Display professional metadata tracking our data engineering stack
%watermark -a "Maykon - AI_Dependency 🤖🧠" -d -u -v -p pandas,numpy,matplotlib,seaborn,scikit-learn,scikit-optimize,jupyterlab,nbconvert,imblearn,xgboost,sweetviz,joblib

Author: Maykon - AI_Dependency 🤖🧠

Last updated: 2026-08-02

Python implementation: CPython
Python version       : 3.14.5
IPython version      : 9.16.0

pandas         : 3.0.5
numpy          : 2.5.1
matplotlib     : 3.11.1
seaborn        : 0.13.2
scikit-learn   : 1.9.0
scikit-optimize: 0.10.2
jupyterlab     : 4.6.2
nbconvert      : 7.17.1
imblearn       : 0.0
xgboost        : 3.3.0
sweetviz       : 2.3.3
joblib         : 1.5.3



In [39]:
# --- AUTOMATED KAGGLE INGESTION ---
# Download the latest version of the specific student behavioral dataset
path = kagglehub.dataset_download("sridipbasu/ai-depndency-career-anxiety-and-student-burnout")
print("🚀 Path to dataset files:", path)

🚀 Path to dataset files: C:\Users\mayki\.cache\kagglehub\datasets\sridipbasu\ai-depndency-career-anxiety-and-student-burnout\versions\1


In [40]:
# --- LOCATING AND READING THE CSV ---
# List out all files inside the downloaded repository path to spot the target file
all_files = os.listdir(path)
print("📂 Files discovered in directory:", all_files)

# Filter out all CSV files dynamically
csv_files = [file for file in all_files if file.endswith('.csv')]

if len(csv_files) == 0:
    raise FileNotFoundError("❌ Critical Error: No CSV files found in the downloaded folder!")
else:
    # Grab the primary CSV file found
    csv_filename = csv_files[0]
    full_csv_path = os.path.join(path, csv_filename)
    print(f"🎯 Target CSV located: {csv_filename}")

📂 Files discovered in directory: ['ai_dependency_career_anxiety_students.csv', 'dataset_metadata.json', 'feature_dictionary.csv']
🎯 Target CSV located: ai_dependency_career_anxiety_students.csv


In [41]:
# Ingest the dataset into a pandas DataFrame
df = pd.read_csv(full_csv_path)
print(f"✅ Dataset successfully loaded! Named 'df', Shape: {df.shape[0]} rows, {df.shape[1]} columns.")

✅ Dataset successfully loaded! Named 'df', Shape: 15000 rows, 30 columns.


In [42]:
# Display the first 5 records to see our column properties and labels
df.head()

,student_id,age,gender,degree_type,stream,year_of_study,college_tier,urban_or_rural,daily_ai_tool_usage_hrs,primary_ai_tools_used,...,daily_study_hours,self_learning_hours_per_week,skill_development_courses_taken,social_media_hrs_per_day,sleep_hours,stress_level,burnout_score,motivation_score,seeks_career_counseling,overall_career_readiness_score
0,STU_00001,25,Male,B.Tech/B.E.,Engineering (Non-CS),3,Tier 3,Rural,0.3,Perplexity,...,3.5,8.7,2,1.5,8.5,1,3,6,0.0,9.09
1,STU_00002,20,Female,B.Tech/B.E.,Engineering (Non-CS),4,Tier 3,Urban,1.9,ChatGPT,...,2.4,8.2,3,4.2,8.3,4,5,3,0.0,6.03
2,STU_00003,25,Female,MBA,CS/IT,1,Tier 3,Urban,3.6,Gemini,...,2.3,16.4,1,2.3,7.0,5,9,3,1.0,6.68
3,STU_00004,23,Male,B.Tech/B.E.,CS/IT,1,Tier 1,Urban,4.1,GitHub Copilot,...,7.7,15.0,3,2.6,7.2,6,6,6,0.0,7.71
4,STU_00005,22,Female,MBA,CS/IT,1,Tier 1,Urban,3.4,ChatGPT,...,2.4,1.1,2,1.9,8.0,3,5,4,0.0,4.64


In [43]:
df.shape # (rows, columns)
print("Number of rows:", df.shape[0])
print("Number of columns:", df.shape[1])

Number of rows: 15000
Number of columns: 30


In [44]:
df.sample(15) # Random 15 rows

,student_id,age,gender,degree_type,stream,year_of_study,college_tier,urban_or_rural,daily_ai_tool_usage_hrs,primary_ai_tools_used,...,daily_study_hours,self_learning_hours_per_week,skill_development_courses_taken,social_media_hrs_per_day,sleep_hours,stress_level,burnout_score,motivation_score,seeks_career_counseling,overall_career_readiness_score
5431,STU_05432,20,Male,MBA,Commerce/Management,1,Tier 1,Urban,0.0,NaN,...,2.2,4.0,1,7.1,4.3,4,4,4,0.0,5.03
1213,STU_01214,21,Female,B.Sc/B.A/B.Com,Commerce/Management,2,Tier 3,Urban,3.3,Perplexity,...,2.2,13.3,2,0.6,5.3,5,6,4,0.0,7.80
3649,STU_03650,20,Male,MBA,CS/IT,1,Tier 1,Urban,3.7,ChatGPT,...,2.4,11.7,1,2.4,7.7,4,6,4,0.0,7.25
11592,STU_11593,23,Male,M.Tech/M.Sc,Arts/Sciences,1,Tier 3,Urban,0.0,NaN,...,4.0,9.1,4,0.0,8.0,5,3,8,0.0,7.48
10205,STU_10206,22,Male,M.Tech/M.Sc,Arts/Sciences,2,Tier 1,Rural,0.0,NaN,...,6.1,4.9,2,0.0,7.5,5,6,6,0.0,6.35
2446,STU_02447,23,Male,B.Sc/B.A/B.Com,CS/IT,1,Tier 3,Urban,2.4,ChatGPT,...,3.3,14.3,6,3.0,8.1,4,6,5,0.0,6.71
3178,STU_03179,20,Male,B.Tech/B.E.,Engineering (Non-CS),1,Tier 3,Rural,1.5,ChatGPT,...,3.2,9.4,3,0.0,8.1,5,4,6,0.0,6.13
1817,STU_01818,20,Female,M.Tech/M.Sc,Commerce/Management,2,Tier 3,Rural,2.0,ChatGPT,...,1.6,2.0,0,2.4,8.5,5,6,3,0.0,5.24
10224,STU_10225,19,Male,M.Tech/M.Sc,Engineering (Non-CS),1,Tier 2,Urban,0.0,NaN,...,3.3,0.8,2,2.3,7.0,7,6,5,0.0,5.41
14679,STU_14680,21,Female,M.Tech/M.Sc,CS/IT,2,Tier 2,Urban,4.2,Perplexity,...,0.3,2.9,0,3.7,7.5,5,7,2,0.0,4.54


In [45]:
df.tail() #Displays the last 5 rows of the DataFrame df.

,student_id,age,gender,degree_type,stream,year_of_study,college_tier,urban_or_rural,daily_ai_tool_usage_hrs,primary_ai_tools_used,...,daily_study_hours,self_learning_hours_per_week,skill_development_courses_taken,social_media_hrs_per_day,sleep_hours,stress_level,burnout_score,motivation_score,seeks_career_counseling,overall_career_readiness_score
14995,STU_14996,21,Female,B.Sc/B.A/B.Com,CS/IT,1,Tier 3,Urban,2.8,ChatGPT,...,0.0,6.2,2,0.2,9.6,4,6,2,0.0,7.02
14996,STU_14997,22,Male,B.Tech/B.E.,Commerce/Management,4,Tier 1,Urban,3.6,ChatGPT,...,6.1,11.7,3,2.9,5.8,5,3,9,1.0,7.31
14997,STU_14998,22,Male,B.Tech/B.E.,CS/IT,2,Tier 1,Rural,1.7,Gemini,...,6.3,17.2,5,0.6,7.4,3,1,9,0.0,8.65
14998,STU_14999,22,Female,MBA,CS/IT,2,Tier 2,Urban,0.0,NaN,...,0.8,0.0,1,2.0,9.1,5,6,2,0.0,4.97
14999,STU_15000,23,Male,M.Tech/M.Sc,Engineering (Non-CS),2,Tier 3,Rural,2.1,ChatGPT,...,2.9,1.0,3,3.9,7.4,7,7,4,0.0,5.28


In [46]:
df.dtypes #Displays the data type of each column in the DataFrame.

student_id                             str
age                                  int64
gender                                 str
degree_type                            str
stream                                 str
year_of_study                        int64
college_tier                           str
urban_or_rural                         str
daily_ai_tool_usage_hrs            float64
primary_ai_tools_used                  str
uses_ai_for_assignments                str
ai_replaces_own_thinking_score       int64
ai_dependency_score                  int64
placement_anxiety_score              int64
fear_of_job_loss_to_ai               int64
career_clarity_score                 int64
internship_experience                int64
weekly_job_application_count         int64
resume_confidence_score              int64
interview_anxiety_score              int64
daily_study_hours                  float64
self_learning_hours_per_week       float64
skill_development_courses_taken      int64
social_medi

In [47]:
df.columns #Returns a list (Index object) containing the names of all columns in the DataFrame.

Index(['student_id', 'age', 'gender', 'degree_type', 'stream', 'year_of_study',
       'college_tier', 'urban_or_rural', 'daily_ai_tool_usage_hrs',
       'primary_ai_tools_used', 'uses_ai_for_assignments',
       'ai_replaces_own_thinking_score', 'ai_dependency_score',
       'placement_anxiety_score', 'fear_of_job_loss_to_ai',
       'career_clarity_score', 'internship_experience',
       'weekly_job_application_count', 'resume_confidence_score',
       'interview_anxiety_score', 'daily_study_hours',
       'self_learning_hours_per_week', 'skill_development_courses_taken',
       'social_media_hrs_per_day', 'sleep_hours', 'stress_level',
       'burnout_score', 'motivation_score', 'seeks_career_counseling',
       'overall_career_readiness_score'],
      dtype='str')

### 🧹 Data Preprocessing & Hygiene
Now that the data is loaded, we begin our cleaning pipeline. 
Our tasks in this phase are:
1. **Find the missing data:** Using the command **df.isna().sum()** 
2. **Drop Identifiers:** Remove `student_id` since it has no predictive or statistical power.
3. **Handle Missing Data:** Clean the columns containing missing values ($NaN$) using imputation strategies:
   * **Categorical Imputation:** Fill high-missingness text columns like `primary_ai_tools_used` with a default placeholder (`'None'`).
   * **Numerical Imputation:** Fill missing numerical features with their **Median** values to keep our distributions stable and unaffected by outliers.

In [48]:
df.isna().sum() # Count missing values per column

student_id                            0
age                                   0
gender                                0
degree_type                           0
stream                                0
year_of_study                         0
college_tier                          0
urban_or_rural                        0
daily_ai_tool_usage_hrs               0
primary_ai_tools_used              3215
uses_ai_for_assignments               0
ai_replaces_own_thinking_score        0
ai_dependency_score                   0
placement_anxiety_score               0
fear_of_job_loss_to_ai                0
career_clarity_score                  0
internship_experience                 0
weekly_job_application_count          0
resume_confidence_score               0
interview_anxiety_score               0
daily_study_hours                     0
self_learning_hours_per_week        233
skill_development_courses_taken       0
social_media_hrs_per_day            210
sleep_hours                         203


In [49]:
df.describe(include='all').T #Generates a complete statistical summary of the DataFrame.

,count,unique,top,freq,mean,std,min,25%,50%,75%,max
student_id,15000,15000,STU_00001,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN
age,15000.0,NaN,NaN,NaN,21.384533,2.005626,18.0,20.0,21.0,22.0,28.0
gender,15000,3,Male,7840,NaN,NaN,NaN,NaN,NaN,NaN,NaN
degree_type,15000,4,B.Tech/B.E.,6700,NaN,NaN,NaN,NaN,NaN,NaN,NaN
stream,15000,4,CS/IT,6091,NaN,NaN,NaN,NaN,NaN,NaN,NaN
year_of_study,15000.0,NaN,NaN,NaN,2.100067,0.98765,1.0,1.0,2.0,3.0,4.0
college_tier,15000,3,Tier 3,7475,NaN,NaN,NaN,NaN,NaN,NaN,NaN
urban_or_rural,15000,2,Urban,10282,NaN,NaN,NaN,NaN,NaN,NaN,NaN
daily_ai_tool_usage_hrs,15000.0,NaN,NaN,NaN,2.013893,1.735977,0.0,0.4,1.8,3.2,8.0
primary_ai_tools_used,11785,5,ChatGPT,5469,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [50]:
df.info() #Displays a summary of the DataFrame structure. 

<class 'pandas.DataFrame'>
RangeIndex: 15000 entries, 0 to 14999
Data columns (total 30 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   student_id                       15000 non-null  str    
 1   age                              15000 non-null  int64  
 2   gender                           15000 non-null  str    
 3   degree_type                      15000 non-null  str    
 4   stream                           15000 non-null  str    
 5   year_of_study                    15000 non-null  int64  
 6   college_tier                     15000 non-null  str    
 7   urban_or_rural                   15000 non-null  str    
 8   daily_ai_tool_usage_hrs          15000 non-null  float64
 9   primary_ai_tools_used            11785 non-null  str    
 10  uses_ai_for_assignments          15000 non-null  str    
 11  ai_replaces_own_thinking_score   15000 non-null  int64  
 12  ai_dependency_score          

In [51]:
df.drop_duplicates(inplace=True) # Remove duplicate rows from the DataFrame df.
print("Number of duplicate rows Now:", df.duplicated().sum()) # After dropping duplicates, check again to confirm that there are no duplicate rows remaining in the DataFrame df.

Number of duplicate rows Now: 0


In [52]:
# ============================================================
# 🗑️ DROP IRRELEVANT COLUMNS
# ============================================================

cols_to_drop = [
    'student_id',        # unique identifier — no predictive power
]

df = df.drop(columns=cols_to_drop, errors='ignore')
print(f"✅ Dropped {len(cols_to_drop)} columns")
print(f"   Remaining columns: {df.shape[1]}")

✅ Dropped 1 columns
   Remaining columns: 29


In [53]:
# ============================================================
# 🧹 COLUMN NAME SANITIZATION
# ============================================================
df.columns = (
    df.columns
    .str.lower()                                        # lowercase everything
    .str.strip()                                        # remove leading/trailing whitespace
    .str.replace(r"[^a-z0-9]+", "_", regex=True)       # replace anything not a letter/number with _
    .str.strip("_")                                     # remove leading/trailing underscores
)

print("✅ Columns sanitized:")
print(df.columns.tolist())

✅ Columns sanitized:
['age', 'gender', 'degree_type', 'stream', 'year_of_study', 'college_tier', 'urban_or_rural', 'daily_ai_tool_usage_hrs', 'primary_ai_tools_used', 'uses_ai_for_assignments', 'ai_replaces_own_thinking_score', 'ai_dependency_score', 'placement_anxiety_score', 'fear_of_job_loss_to_ai', 'career_clarity_score', 'internship_experience', 'weekly_job_application_count', 'resume_confidence_score', 'interview_anxiety_score', 'daily_study_hours', 'self_learning_hours_per_week', 'skill_development_courses_taken', 'social_media_hrs_per_day', 'sleep_hours', 'stress_level', 'burnout_score', 'motivation_score', 'seeks_career_counseling', 'overall_career_readiness_score']


#### 🧠 Create the profiling report

In [54]:
# create the instance of the Sweetviz report
report = sv.analyze(df)

# Save the report as an HTML file
report.show_html("Report_AI_dependency.html")

# Success message
print("✅ Success! The Sweetviz report was created successfully.")
print("📄 Report saved as: Report_AI_dependency.html") 

Done! Use 'show' commands to display/save.   |██████████| [100%]   00:00 -> (00:00 left)

Report Report_AI_dependency.html was generated! NOTEBOOK/COLAB USERS: the web browser MAY not pop up, regardless, the report IS saved in your notebook/colab files.
✅ Success! The Sweetviz report was created successfully.
📄 Report saved as: Report_AI_dependency.html


#### 🎯 Defining the data between Features (X) and Target (y)

In supervised machine learning, we divide the dataset into:

- **Features (`X`)** → the input variables used to make the prediction.
- **Target variable (`y`)** → the outcome we want to predict.

For the AI_Dependency dataset:

#### Creating the Target Variable

The raw `ai_dependency_score` runs from 1 to 10 — a continuous 
ordinal scale that we convert into 3 meaningful behavioral classes:

| Score | Label | Meaning |
|---|---|---|
| 1–3 | **Low** | Minimal AI reliance — student works mostly independently |
| 4–6 | **Medium** | Moderate daily use — AI as a tool, not a crutch |
| 7–10 | **High** | Heavy dependency — AI replaces independent thinking |

This transforms our problem into a **3-class classification task**.

In [55]:
# ============================================================
# 🎯 CREATE TARGET VARIABLE
# ============================================================

def categorize_dependency(score):
    if score <= 3:
        return 'Low'
    elif score <= 6:
        return 'Medium'
    else:
        return 'High'

df['ai_dependency_label'] = df['ai_dependency_score'].apply(categorize_dependency)

print("✅ Target variable created!")
print()
print(df['ai_dependency_label'].value_counts())
print()
print("Class distribution (%):")
print((df['ai_dependency_label'].value_counts(normalize=True) * 100).round(1))

✅ Target variable created!

ai_dependency_label
Medium    6445
Low       5407
High      3148
Name: count, dtype: int64

Class distribution (%):
ai_dependency_label
Medium    43.0
Low       36.0
High      21.0
Name: proportion, dtype: float64


In [56]:
# ============================================================
# 🗑️ DROP LEAKAGE COLUMNS
# ============================================================

cols_to_drop = [
    'ai_dependency_score',             # raw score — replaced by label
    'ai_replaces_own_thinking_score',  # directly measures same thing — leakage
]

df = df.drop(columns=cols_to_drop, errors='ignore')
print(f"✅ Dropped: {cols_to_drop}")
print(f"   Remaining columns: {df.shape[1]}")

✅ Dropped: ['ai_dependency_score', 'ai_replaces_own_thinking_score']
   Remaining columns: 28


In [57]:
# ============================================================
# ✂️ TRAIN / TEST SPLIT
# ============================================================

X = df.drop(columns=['ai_dependency_label'])
y = df['ai_dependency_label']

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size   = 0.2,
    random_state= 42,
    stratify    = y   # ✅ preserves Low/Medium/High balance
)

print(f"✅ Split complete!")
print(f"   X_train : {X_train.shape}")
print(f"   X_test  : {X_test.shape}")
print()
print("Training class distribution:")
print(y_train.value_counts())
print()
print("Test class distribution:")
print(y_test.value_counts())

✅ Split complete!
   X_train : (12000, 27)
   X_test  : (3000, 27)

Training class distribution:
ai_dependency_label
Medium    5156
Low       4326
High      2518
Name: count, dtype: int64

Test class distribution:
ai_dependency_label
Medium    1289
Low       1081
High       630
Name: count, dtype: int64


#### 🔢 Defining Numeric and Categorical Columns

We separate our features into two groups so the preprocessing 
pipeline can apply the right transformation to each:

- **Numeric** → impute missing values with median, then scale
- **Categorical** → impute missing values with most frequent, then encode

This separation is the foundation of the `ColumnTransformer` pipeline.

In [58]:
print("----" * 12)
print(X_train.dtypes)
print("----" * 12)
print(f"Total columns: {X_train.shape[1]} 👈")

------------------------------------------------
age                                  int64
gender                                 str
degree_type                            str
stream                                 str
year_of_study                        int64
college_tier                           str
urban_or_rural                         str
daily_ai_tool_usage_hrs            float64
primary_ai_tools_used                  str
uses_ai_for_assignments                str
placement_anxiety_score              int64
fear_of_job_loss_to_ai               int64
career_clarity_score                 int64
internship_experience                int64
weekly_job_application_count         int64
resume_confidence_score              int64
interview_anxiety_score              int64
daily_study_hours                  float64
self_learning_hours_per_week       float64
skill_development_courses_taken      int64
social_media_hrs_per_day           float64
sleep_hours                        float64
stres

In [59]:
# ============================================================
# DEFINE FEATURE TYPES
# ============================================================

numeric_features = [
    'age',
    'daily_ai_tool_usage_hrs',
    'placement_anxiety_score',
    'fear_of_job_loss_to_ai',
    'career_clarity_score',
    'internship_experience',
    'weekly_job_application_count',
    'resume_confidence_score',
    'interview_anxiety_score',
    'daily_study_hours',
    'self_learning_hours_per_week',
    'skill_development_courses_taken',
    'social_media_hrs_per_day',
    'sleep_hours',
    'stress_level',
    'burnout_score',
    'motivation_score',
    'seeks_career_counseling',
    'overall_career_readiness_score',
]

categorical_features = [
    'gender',
    'degree_type',
    'stream',
    'year_of_study',
    'college_tier',
    'urban_or_rural',
    'primary_ai_tools_used',
    'uses_ai_for_assignments',
]

# Sanity check — no overlaps, no missing columns
all_features = numeric_features + categorical_features
missing      = [col for col in X_train.columns if col not in all_features]
overlap      = [col for col in numeric_features if col in categorical_features]

print(f"✅ Numeric features    : {len(numeric_features)}")
print(f"✅ Categorical features: {len(categorical_features)}")
print(f"✅ Total defined       : {len(all_features)}")
print(f"✅ Total in X_train    : {X_train.shape[1]}")
print()
if missing:
    print(f"⚠️🚨 Missing from lists  : {missing}")
else:
    print("✅ All columns accounted for!")
if overlap:
    print(f"❌ Overlap detected    : {overlap}")
else:
    print("✅ No overlaps detected!")

✅ Numeric features    : 19
✅ Categorical features: 8
✅ Total defined       : 27
✅ Total in X_train    : 27

✅ All columns accounted for!
✅ No overlaps detected!


#### 🔧 Building the Preprocessing Pipeline

We build two separate sub-pipelines — one for numeric features, 
one for categorical — then combine them using a `ColumnTransformer`.

Each pipeline handles missing values first, then transforms the data:
- **Numeric** → median imputation → standard scaling
- **Categorical** → most frequent imputation → one-hot encoding

Everything lives inside the pipeline — no manual transformations 
on `X_train` or `X_test`. This guarantees zero data leakage across 
all cross-validation folds.

In [60]:
# ============================================================
# 🔧 PREPROCESSING PIPELINE
# ============================================================

# --- Numeric: impute with median, then scale ---
numeric_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='median')),
    ('scaler',  StandardScaler())
])

# --- Categorical: impute with most frequent, then encode ---
categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('onehot',  OneHotEncoder(handle_unknown='ignore', sparse_output=False))
])

# --- Combine both into one preprocessor ---
preprocessor = ColumnTransformer(transformers=[
    ('num', numeric_transformer, numeric_features),
    ('cat', categorical_transformer, categorical_features)
],
    remainder='drop'
)

print("✅ Preprocessing pipeline built!")
print(f"   Numeric transformations  : median imputation → standard scaling")
print(f"   Categorical transformations: most frequent imputation → one-hot encoding")

✅ Preprocessing pipeline built!
   Numeric transformations  : median imputation → standard scaling
   Categorical transformations: most frequent imputation → one-hot encoding


#### 🔁 Cross-Validation Strategy

We use `StratifiedKFold` with 5 folds to evaluate all models consistently.

`StratifiedKFold` is essential here because our classes are mildly 
imbalanced (Low 36%, Medium 43%, High 21%) — it guarantees each fold 
preserves the same class proportions as the full training set.

Every model — including the baseline — is evaluated with the same 
CV strategy for a fair, apples-to-apples comparison.

In [61]:
# ============================================================
# 🔁 CROSS-VALIDATION STRATEGY
# ============================================================

cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

print("✅ Cross-validation strategy defined!")
print(f"   Type     : StratifiedKFold")
print(f"   Folds    : 5")
print(f"   Shuffle  : True")
print(f"   Seed     : 42")

✅ Cross-validation strategy defined!
   Type     : StratifiedKFold
   Folds    : 5
   Shuffle  : True
   Seed     : 42


#### 🏁 Baseline Model — DummyClassifier

Before training any real model, we establish a baseline using 
`DummyClassifier` with `strategy='most_frequent'` — it always 
predicts the majority class (Medium) regardless of the input.

Any real model that can't beat this baseline is useless.
This is our minimum bar to clear.

In [62]:
# ============================================================
# 🏁 BASELINE — DummyClassifier
# ============================================================

dummy_pipeline = Pipeline(steps=[('preprocessor', preprocessor), ('model', DummyClassifier(strategy='most_frequent'))])

dummy_accuracy = cross_val_score(dummy_pipeline, X_train, y_train, cv=cv, scoring='accuracy')
dummy_f1       = cross_val_score(dummy_pipeline, X_train, y_train, cv=cv, scoring='f1_weighted')

print("=" * 55)
print("🏁 BASELINE (DummyClassifier — most frequent)")
print("=" * 55)
print(f"Accuracy : {dummy_accuracy.mean():.4f} (+/- {dummy_accuracy.std():.4f})")
print(f"F1 Score : {dummy_f1.mean():.4f} (+/- {dummy_f1.std():.4f})")
print("=" * 55)

🏁 BASELINE (DummyClassifier — most frequent)
Accuracy : 0.4297 (+/- 0.0002)
F1 Score : 0.2583 (+/- 0.0002)


What these numbers mean

**Accuracy: 0.4297 (43%)**
By always predicting "Medium" — the most frequent class — the model is right 43% of the time. That's our minimum bar to beat.

**F1 Score: 0.2583 (26%)**
F1 is much lower than accuracy — this confirms the class imbalance. A model that ignores Low and High classes completely gets punished heavily on F1. This is why F1 is more honest than accuracy here.

+/- 0.0002
Extremely stable across folds — expected for a dummy model that always predicts the same thing.

#### 🤖 Model Comparison

We train and evaluate 3 models using the same `StratifiedKFold` 
strategy as the baseline — ensuring a fair comparison across all models.

Every model is wrapped in a full pipeline (preprocessor + model) 
so preprocessing is learned only from training folds — no leakage.

#### 📈 Logistic Regression
The first real model you test after the baseline.
It's a simple and interpretable algorithm that tries to find a straight line (or boundary) that separates survivors from non-survivors based on your features. It's not the most powerful model, but it's a great starting point because it's fast, easy to understand, and gives you a reference score to beat with more complex models like Random Forest.

In [63]:
# ============================================================
# 📊 MODEL 1 — Logistic Regression
# ============================================================

lr_pipeline = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('model', LogisticRegression(max_iter=1000, random_state=42))
])

lr_accuracy = cross_val_score(lr_pipeline, X_train, y_train, cv=cv, scoring='accuracy')
lr_f1       = cross_val_score(lr_pipeline, X_train, y_train, cv=cv, scoring='f1_weighted')

print("=" * 55)
print("📊 MODEL 1 — Logistic Regression")
print("=" * 55)
print(f"Accuracy : {lr_accuracy.mean():.4f} (+/- {lr_accuracy.std():.4f})")
print(f"F1 Score : {lr_f1.mean():.4f} (+/- {lr_f1.std():.4f})")
print("=" * 55)

📊 MODEL 1 — Logistic Regression
Accuracy : 0.6708 (+/- 0.0060)
F1 Score : 0.6707 (+/- 0.0060)


#### 🌲 Random Forest
Unlike Logistic Regression which draws a single straight boundary, Random Forest builds hundreds of decision trees and combines their votes to make a final prediction. This makes it much better at capturing complex patterns in the data.

In [64]:
# ============================================================
# 📊 MODEL 2 — Random Forest
# ============================================================

rf_pipeline = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('model', RandomForestClassifier(random_state=42))
])

rf_accuracy = cross_val_score(rf_pipeline, X_train, y_train, cv=cv, scoring='accuracy')
rf_f1       = cross_val_score(rf_pipeline, X_train, y_train, cv=cv, scoring='f1_weighted')

print("=" * 55)
print("📊 MODEL 2 — Random Forest")
print("=" * 55)
print(f"Accuracy : {rf_accuracy.mean():.4f} (+/- {rf_accuracy.std():.4f})")
print(f"F1 Score : {rf_f1.mean():.4f} (+/- {rf_f1.std():.4f})")
print("=" * 55)

📊 MODEL 2 — Random Forest
Accuracy : 0.6505 (+/- 0.0120)
F1 Score : 0.6500 (+/- 0.0124)


#### ⚡ Gradient Boosting

Gradient Boosting (using Scikit-Learn's GradientBoostingClassifier).

While Random Forest builds many trees at the same time independently (bagging), Gradient Boosting builds trees one after the other (boosting). Each new tree is specifically designed to focus on and correct the mistakes made by the previous trees.

Why we love it: It is often the most accurate algorithm for tabular data, frequently winning Kaggle competitions. It is incredibly sharp at picking up subtle, complex patterns.

The Catch: Because it builds trees sequentially, it takes longer to train than a Random Forest and can overfit if you let it build too many trees.

In [65]:
# ============================================================
# 📊 MODEL 3 — Gradient Boosting
# ============================================================

gb_pipeline = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('model', GradientBoostingClassifier(random_state=42))
])

gb_accuracy = cross_val_score(gb_pipeline, X_train, y_train, cv=cv, scoring='accuracy')
gb_f1       = cross_val_score(gb_pipeline, X_train, y_train, cv=cv, scoring='f1_weighted')

print("=" * 55)
print("📊 MODEL 3 — Gradient Boosting")
print("=" * 55)
print(f"Accuracy : {gb_accuracy.mean():.4f} (+/- {gb_accuracy.std():.4f})")
print(f"F1 Score : {gb_f1.mean():.4f} (+/- {gb_f1.std():.4f})")
print("=" * 55)

📊 MODEL 3 — Gradient Boosting
Accuracy : 0.6596 (+/- 0.0095)
F1 Score : 0.6593 (+/- 0.0098)


##### 🏆 MODEL COMPARISON

In [66]:
print("=" * 65)
print(f"{'Model':<25} | {'Accuracy':>10} | {'F1 Weighted':>12}")
print("=" * 65)
print(f"{'Baseline':<25} | {dummy_accuracy.mean():>10.4f} | {dummy_f1.mean():>12.4f}")
print(f"{'Logistic Regression':<25} | {lr_accuracy.mean():>10.4f} | {lr_f1.mean():>12.4f}")
print(f"{'Random Forest':<25} | {rf_accuracy.mean():>10.4f} | {rf_f1.mean():>12.4f}")
print(f"{'Gradient Boosting':<25} | {gb_accuracy.mean():>10.4f} | {gb_f1.mean():>12.4f}")
print("=" * 65)

# Automatic winner detection
results = {
    'Logistic Regression' : lr_f1.mean(),
    'Random Forest'       : rf_f1.mean(),
    'Gradient Boosting'   : gb_f1.mean()
}

winner = max(results, key=results.get)
print()
print(f"🏆 Winner: {winner} (F1: {results[winner]:.4f})")

Model                     |   Accuracy |  F1 Weighted
Baseline                  |     0.4297 |       0.2583
Logistic Regression       |     0.6708 |       0.6707
Random Forest             |     0.6505 |       0.6500
Gradient Boosting         |     0.6596 |       0.6593

🏆 Winner: Logistic Regression (F1: 0.6707)


**What this tells:**

1. All models crushed the baseline — jumping from 0.26 to 0.67 F1 means the features are genuinely predictive of AI dependency. ✅

2. Logistic Regression won again — just like the house price project. This tells you the relationship between features and AI dependency is largely linear. Complex models aren't finding extra patterns.

3. All three models are very close — only 0.02 F1 separates first from last. This means the dataset has a natural performance ceiling around 67% that complexity alone won't break.

4. Accuracy ≈ F1 — the gap between accuracy and F1 is tiny for all models. The mild class imbalance isn't severely hurting performance — good news, no SMOTE needed yet.

**This tells us exactly how the model performs on each class individually**

In [67]:
# ============================================================
# 🔍 PER-CLASS PERFORMANCE CHECK
# ============================================================
y_pred_cv = cross_val_predict(lr_pipeline, X_train, y_train, cv=cv) 
print(classification_report(y_train, y_pred_cv, target_names=['High', 'Low', 'Medium']))

              precision    recall  f1-score   support

        High       0.71      0.61      0.66      2518
         Low       0.71      0.74      0.72      4326
      Medium       0.62      0.65      0.63      5156

    accuracy                           0.67     12000
   macro avg       0.68      0.66      0.67     12000
weighted avg       0.67      0.67      0.67     12000



#### **Key insights:**

**1. High class F1 = 0.66 ✅**
This is above our 0.60 threshold — no SMOTE needed, no class_weight='balanced' needed. The model is handling the minority class well despite having fewer samples.

**2. Low class is the strongest (F1 = 0.72)**
Students with minimal AI dependency have the most distinct behavioral pattern — easiest for the model to identify.

**3. Medium class is the weakest (F1 = 0.63)**
This makes intuitive sense — Medium students sit between Low and High, making them harder to distinguish. Their behavior overlaps with both extremes.

**4. High recall is low (0.61)**
The model misses 39% of truly High dependency students — predicting them as Medium instead. This is the main area to improve during tuning.

#### 🔧 Hyperparameter Tuning — Logistic Regression (Winner)

Logistic Regression won the model comparison with the highest F1 score.
We now tune its key parameters using `BayesSearchCV` — a smarter 
search strategy that learns from each trial instead of testing blindly.

Key parameters to tune:
- **`C`** → regularization strength (lower = more penalty)
- **`penalty`** → L1 (feature selection) or L2 (shrinkage)
- **`solver`** → optimization algorithm

In [73]:
# ============================================================
# 🔧 HYPERPARAMETER TUNING — Logistic Regression
# ============================================================
param_space = {
    'model__C'       : Real(0.01, 100.0, prior='log-uniform'),
    'model__l1_ratio': Real(0.0, 1.0)  # 0.0 = pure L2, 1.0 = pure L1
}

lr_tuned_pipeline = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('model', LogisticRegression(
        solver     = 'saga',
        penalty    = 'elasticnet',  # ✅ supports l1_ratio
        max_iter   = 1000,
        random_state = 42
    ))
])

bayes_search = BayesSearchCV(
    estimator     = lr_tuned_pipeline,
    search_spaces = param_space,
    n_iter        = 30,
    cv            = cv,
    scoring       = 'f1_weighted',
    n_jobs        = -1,
    random_state  = 42
)

bayes_search.fit(X_train, y_train)

print(f"✅ Best params : {dict(bayes_search.best_params_)}")
print(f"✅ Best F1     : {bayes_search.best_score_:.4f}")

c:\Users\mayki\OneDrive\Desktop\Projects\AI_Dependency\.venv\Lib\site-packages\sklearn\linear_model\_logistic.py:1403: FutureWarning: 'penalty' was deprecated in version 1.8 and will be removed in 1.10. To avoid this warning, leave 'penalty' set to its default value and use 'l1_ratio' or 'C' instead. Use l1_ratio=0 instead of penalty='l2', l1_ratio=1 instead of penalty='l1', l1_ratio set to a float between 0 and 1 instead of penalty='elasticnet', and C=np.inf instead of penalty=None.
  warnings.warn(


✅ Best params : {'model__C': 0.06799214551856791, 'model__l1_ratio': 0.6291462266456603}
✅ Best F1     : 0.6724
